From a0d28a32698fbe52f64faada38abb5325a71f7eb Mon Sep 17 00:00:00 2001
From: Radim Krcmar <rkrcmar@redhat.com>
Date: Wed, 12 Nov 2014 16:32:02 -0500
Subject: [virt] kvm: trace kvm_ple_window grow/shrink

Message-id: <1415809922-10128-6-git-send-email-rkrcmar@redhat.com>
Patchwork-id: 99849
O-Subject: [RHEL7.1 kernel PATCH 5/5] KVM: trace kvm_ple_window grow/shrink
Bugzilla: 1163296
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Andrew Jones <drjones@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

Tracepoint for dynamic PLE window, fired on every potential change.

Signed-off-by: Radim Krcmar <rkrcmar@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 7b46268d29543e313e731606d845e65c17f232e4)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/trace.h b/arch/x86/kvm/trace.h
index 33574c9..23246cf 100644
--- a/arch/x86/kvm/trace.h
+++ b/arch/x86/kvm/trace.h
@@ -848,6 +848,36 @@ TRACE_EVENT(kvm_track_tsc,
     __print_symbolic(__entry->host_clock, host_clocks))
 );
 
+TRACE_EVENT(kvm_ple_window,
+ TP_PROTO(bool grow, unsigned int vcpu_id, int new, int old),
+ TP_ARGS(grow, vcpu_id, new, old),
+
+ TP_STRUCT__entry(
+  __field(                bool,      grow         )
+  __field(        unsigned int,   vcpu_id         )
+  __field(                 int,       new         )
+  __field(                 int,       old         )
+ ),
+
+ TP_fast_assign(
+  __entry->grow           = grow;
+  __entry->vcpu_id        = vcpu_id;
+  __entry->new            = new;
+  __entry->old            = old;
+ ),
+
+ TP_printk("vcpu %u: ple_window %d (%s %d)",
+           __entry->vcpu_id,
+           __entry->new,
+           __entry->grow ? "grow" : "shrink",
+           __entry->old)
+);
+
+#define trace_kvm_ple_window_grow(vcpu_id, new, old) \
+ trace_kvm_ple_window(true, vcpu_id, new, old)
+#define trace_kvm_ple_window_shrink(vcpu_id, new, old) \
+ trace_kvm_ple_window(false, vcpu_id, new, old)
+
 #endif /* CONFIG_X86_64 */
 
 #endif /* _TRACE_KVM_H */
diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index e94161f..b782d06 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -5729,6 +5729,8 @@ static void grow_ple_window(struct kvm_vcpu *vcpu)
 
  if (vmx->ple_window != old)
   vmx->ple_window_dirty = true;
+
+ trace_kvm_ple_window_grow(vcpu->vcpu_id, vmx->ple_window, old);
 }
 
 static void shrink_ple_window(struct kvm_vcpu *vcpu)
@@ -5741,6 +5743,8 @@ static void shrink_ple_window(struct kvm_vcpu *vcpu)
 
  if (vmx->ple_window != old)
   vmx->ple_window_dirty = true;
+
+ trace_kvm_ple_window_shrink(vcpu->vcpu_id, vmx->ple_window, old);
 }
 
 /*
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index a518ce5..c33d648 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -7615,3 +7615,4 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(kvm_invlpga);
 EXPORT_TRACEPOINT_SYMBOL_GPL(kvm_skinit);
 EXPORT_TRACEPOINT_SYMBOL_GPL(kvm_nested_intercepts);
 EXPORT_TRACEPOINT_SYMBOL_GPL(kvm_write_tsc_offset);
+EXPORT_TRACEPOINT_SYMBOL_GPL(kvm_ple_window);
-- 
1.7.1